client/header/utils.js

```diff
@@ -0,0 +1,99 @@
+/**
+ * External dependencies
+ */
+import { Slot, Fill } from '@wordpress/components';
+import { cloneElement } from '@wordpress/element';
+
+/**
+ * Ordered header item.
+ *
+ * @param {Node} children - Node children.
+ * @param {number} order - Node order.
+ * @param {Array} props - Fill props.
+ * @return {Node} Node.
+ */
+const createOrderedChildren = ( children, order, props ) => {
+	return typeof children === 'function'
+		? cloneElement( children( props ), { order } )
+		: cloneElement( children, { ...props, order } );
+};
+
+/**
+ * Create a Fill for extensions to add items to the WooCommerce Admin header.
+ *
+ * @slotFill WooHeaderItem
+ * @example
+ * const MyHeaderItem = () => (
+ * <WooHeaderItem>My header item</WooHeaderItem>
+ * );
+ *
+ * registerPlugin( 'my-extension', {
+ * render: MyHeaderItem,
+ * scope: 'woocommerce-admin',
+ * } );
+ * @param {Object} param0
+ * @param {Array} param0.children - Node children.
+ * @param {Array} param0.order - Node order.
+ */
+export const WooHeaderItem = ( { children, order = 1 } ) => {
+	const suffix = order >= 0 ? 'after' : 'before';
+
+	return (
+		<Fill name={ `woocommerce_header_item_${ suffix }` }>
+			{ createOrderedChildren.bind( null, children, order ) }
```

Comment:
Nice work, I would of never thought about using `bind`. Having said that, so doing the code below doesn't work? I might be missing something.
```
<Fill name={ `woocommerce_header_item_${ suffix }` }>
  { ( fillProps ) => {
      return createOrderedChildren( children, order, {
	 ...fillProps,
	...props,
    } );
  } }
</Fill>
```

Reply:
I hadn't tried that after creating `createOrderedChildren`, but apparently it does work as long as we are manually cloning the children afterwards.  Update in https://github.com/woocommerce/woocommerce-admin/pull/7805/commits/9ad113680338f96e8a538e74a7ec4f71cc7b6bb1